### PR TITLE
Add Fetch and non-ADS mode to the snapshot cache

### PR DESCRIPTION
**Pull request by kyessenov**
Subset of #43 to ease the reviewing process.

- implement Fetch methods on the cache
- refactor the cache interface to use less custom structures
- expose status information from the cache

cc @junr03

### CODE REVIEW

File: pkg/cache/cache.go [hunk 1]
```diff
@@ -16,79 +16,55 @@
...
+	// CreateWatch returns a new open watch from a non-empty request.
+	//
+	// Value channel produces requested resources, once they are available.  If
+	// the channel is closed prior to cancellation of the watch, an unrecoverable
+	// error has occurred in the producer, and the consumer should close the
+	// corresponding stream.
+	//
+	// Cancel is an optional function to release resources in the producer. If
+	// provided, the consumer may call this function multiple times.
+	CreateWatch(Request) (value chan Response, cancel func())
```
Comment (kyessenov) [inline pkg/cache/cache.go]: This is just de-structuring the Watch to make it easier to follow the channels.

File: pkg/cache/resource.go [hunk 2]
```diff
@@ -76,3 +72,45 @@ func GetResourceName(xds proto.Message) string {
 		return ""
 	}
 }
+
+// GetResourceReferences returns the names for dependent resources (EDS cluster
+// names for CDS, RDS routes names for LDS).
+func GetResourceReferences(resources ...Resource) map[string]bool {
+	out := make(map[string]bool)
+	for _, res := range resources {
+		if res != nil {
```
Comment (taion809) [inline pkg/cache/resource.go]: i would change these conditionals from

```go
if res != nil {}
```

to

```go
if res == nil { continue }
```

Specifically to reduce the depth of this branch and make it easier to read.

File: pkg/cache/resource.go [hunk 3]
```diff
@@ -76,3 +72,45 @@ func GetResourceName(xds proto.Message) string {
...
+						out[v.Name] = true
+					}
+				}
+			case *v2.RouteConfiguration:
+				// no dependencies
+			case *v2.Listener:
+				// extract route configuration names from HTTP connection manager
+				for _, chain := range v.FilterChains {
+					for _, filter := range chain.Filters {
+						if filter.Name == HTTPConnectionManager {
```
Comment (taion809) [inline pkg/cache/resource.go]: same comment as L81

File: pkg/cache/resource.go (hunk 2 above)
Comment (kyessenov) [reply pkg/cache/resource.go]: Done

File: pkg/cache/resource.go (hunk 3 above)
Comment (kyessenov) [reply pkg/cache/resource.go]: Done

File: pkg/cache/simple.go [hunk 4]
```diff
@@ -15,187 +15,278 @@
...
-// SimpleCache can also be used as a config cache for distinct xDS requests.
-// The snapshots are required to contain only the responses for the particular
-// type of the xDS service that the cache serves. Synchronization of multiple
-// caches for different response types is left to the configuration producer.
-type SimpleCache struct {
-	// snapshots are cached resources
-	snapshots map[Key]Snapshot
+// SnapshotCache can operate as a REST or regular xDS backend. The snapshot
+// can be partial, e.g. only include RDS or EDS resources.
+type SnapshotCache struct {
```
[the remark this answers is not in the document]
Comment (kyessenov) [reply pkg/cache/simple.go]: Done.

File: pkg/cache/simple.go [hunk 5]
```diff
@@ -15,187 +15,278 @@
...
+	return value, nil
 }
 
-// Respond to a watch with the snapshot value. The value channel should have capacity not to block.
-// TODO(kuat) this responds immediately and can cause the remote node to spin if it consistently fails to ACK the update
-func respond(watch Watch, snapshot Snapshot, group Key) {
-	typ := watch.Type
-	resources := snapshot.resources[typ]
-	version := snapshot.version
+func (cache *SnapshotCache) freshWatchID() int64 {
```
[the remark this answers is not in the document]
Comment (kyessenov) [reply pkg/cache/simple.go]: Renamed.

File: pkg/cache/simple.go [hunk 6]
```diff
@@ -15,187 +15,278 @@
...
+			info.mu.Lock()
+			delete(info.watches, watchID)
+			info.mu.Unlock()
+		}
+	}
+}
 
-	// the request names must match the snapshot names
+// Respond to a watch with the snapshot value. The value channel should have capacity not to block.
+// TODO(kuat) this responds immediately and can cause the remote node to spin if it consistently fails to ACK the update
```
[the remark this answers is not in the document]
Comment (kyessenov) [reply pkg/cache/simple.go]: Done https://github.com/envoyproxy/go-control-plane/issues/46

File: pkg/cache/status.go [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+// NodeHash computes string identifiers for Envoy nodes.
+type NodeHash interface {
+	// ID function defines a unique string identifier for the remote Envoy node.
+	ID(node *core.Node) string
+}
+
+// StatusInfo tracks the server state for the remote Envoy node.
+// Not all fields are used by all cache implementations.
+type StatusInfo struct {
```
[the remark this answers is not in the document]
Comment (kyessenov) [reply pkg/cache/status.go]: Done

File: pkg/server/server.go [hunk 8]
```diff
@@ -17,60 +17,56 @@ package server
...
-func NewServer(config cache.ConfigWatcher) Server {
-	return &server{config: config}
+// NewServer creates handlers from a config watcher and an optional logger.
+func NewServer(config cache.Cache, callbacks Callbacks) Server {
+	return &server{config: config, callbacks: callbacks}
 }
 
 type server struct {
-	config cache.ConfigWatcher
+	config    cache.Cache
```
[the remark this answers is not in the document]
Comment (kyessenov) [reply pkg/server/server.go]: Done

File: pkg/server/server.go [hunk 9]
```diff
@@ -258,37 +296,69 @@ func (s *server) handler(stream stream, typeURL string) error {
...
 
 func (s *server) StreamListeners(stream v2.ListenerDiscoveryService_StreamListenersServer) error {
-	return s.handler(stream, ListenerType)
+	return s.handler(stream, cache.ListenerType)
+}
+
+// Fetch is the universal fetch method.
+func (s *server) Fetch(ctx context.Context, req *v2.DiscoveryRequest) (*v2.DiscoveryResponse, error) {
+	if s.callbacks != nil {
+		s.callbacks.OnFetchRequest(req)
```
[the remark this answers is not in the document]
Comment (kyessenov) [reply pkg/server/server.go]: Yes, the callback can write to the cache here.

File: pkg/server/server.go [hunk 10]
```diff
@@ -17,60 +17,56 @@ package server
...
 	v2.ListenerDiscoveryServiceServer
 	discovery.AggregatedDiscoveryServiceServer
+
+	// Fetch is the universal fetch method.
+	Fetch(context.Context, *v2.DiscoveryRequest) (*v2.DiscoveryResponse, error)
+}
+
+// Callbacks is a collection of callbacks inserted into the server operation.
+type Callbacks interface {
+	// OnStreamOpen is called once an xDS stream is open with a stream ID and the type URL (or "" for ADS).
```
[the remark this answers is not in the document]
Comment (kyessenov) [reply pkg/server/server.go]: Added a top-level comment.

File: pkg/cache/resource.go [hunk 11]
```diff
@@ -76,3 +72,46 @@ func GetResourceName(xds proto.Message) string {
...
+			// for EDS type, use cluster name or ServiceName override
+			if v.Type == v2.Cluster_EDS {
+				if v.EdsClusterConfig != nil && v.EdsClusterConfig.ServiceName != "" {
+					out[v.EdsClusterConfig.ServiceName] = true
+				} else {
+					out[v.Name] = true
+				}
+			}
+		case *v2.RouteConfiguration:
+			// no dependencies
```
Comment (rshriram) [inline pkg/cache/resource.go]: why? Route depends on the clusters to work properly..

Comment (kyessenov) [reply pkg/cache/resource.go]: It does not depend in the ADS sense. Envoy queries the entire set of clusters, not individual ones by name. It's valid to send a route with no cluster, or to send more clusters than used.

Comment (kyessenov) [reply pkg/cache/resource.go]: Good point though -- expanded the comment in the code.

File: pkg/test/server.go [hunk 12]
```diff
@@ -43,9 +42,9 @@ const (
 type Hasher struct {
 }
 
-// Hash function that always returns same value.
-func (h Hasher) Hash(*core.Node) (cache.Key, error) {
-	return cache.Key(node), nil
+// ID function that always returns the same value.
+func (h Hasher) ID(*core.Node) string {
```
Comment (rshriram) [inline pkg/test/server.go]: NodeId

File: pkg/test/server.go [hunk 13]
```diff
@@ -43,9 +42,9 @@ const (
...
+	return node
```
Comment (rshriram) [inline pkg/test/server.go]: Also need a callback to indicate that a new Envoy has connected to the server or an envoy has disconnected from the server. And a callback that indicates whether the snapshot has been fully transmitted and acked by envoy.

Comment (kyessenov) [reply pkg/test/server.go]: There is a callback for various way Envoy connects to the server. There's also a callback for stream disconnect and fetch response. Snapshot transmission callback is WIP.

File: pkg/test/server.go (hunk 12 above)
Comment (kyessenov) [reply pkg/test/server.go]: Is there something else that needs to be IDed except nodes?

**Comment by kyessenov** [comment]
@junr03 let me know if you want to look at it. I will have followups as well from #43

**Pull request merged**